src/emailjs-tcp-socket.js

```diff
@@ -619,8 +662,16 @@
                 // process the data available on the socket
                 self._onData(readInfo.data);
 
-                // queue the next read
-                self._readLegacySocket();
+                // Queue the next read.
+                // If a STARTTLS handshake might be upcoming, postpone this onto
+                // the task queue so the IMAP client has a chance to call upgradeToSecure;
+                // without this, we might eat the beginning of the handshake.
+                // If we are already secure, just call it (for performance).
+                if (self.ssl) {     // are we secure yet?
+                    self._readLegacySocket();
+                } else {
+                    setZeroTimeout(self._readLegacySocket.bind(self));
```

Comment:
Ok, your second link seems to say that promises can also use macrotasks. However, if Jake Archibald says that they always use microtasks, I'm inclined to trust him. Most importantly, we don't worry about portability in this code since it will only ever run in Chrome, and Chrome definitely uses microtasks for its promises. So I still believe this is solid.

I'll also say that I have been running this code for a couple of weeks and it has been rock solid. I haven't had a single of the net::ERR_INVALID_ARGUMENT errors that were frequent before. Anecdotal, I know.